### PR TITLE
Design to Fail

### PR DESCRIPTION
**Pull request by purquhart**
Added functionality to permit service retries upon receiving a status code 500+.

Usage: Supply optional internalServiceErrorRetries and internalServerErrorRetryInterval parameters during instantiation of Configuration class.

Default values (retry interval based on milliseconds):
internalServiceErrorRetries: 1
internalServiceErrorRetryInterval: 100

### CODE REVIEW

**Comment by hajoeichler** [comment]
Those number should not be hardcoded, but configurable.

**Comment by hajoeichler** [comment]
Could you not abstract the actual call so you don't have to repeat the implementation 5 times.
Have one call method that can be configured to be POST, DELETE etc and have am optional wrapper for repetition.

**Comment by hajoeichler** [comment]
There should be a max retry < 10 (better 3 or 5).
This constraint should be checked in the code.

File: commercetools.NET/Common/Client.cs [hunk 1]
```diff
@@ -277,23 +317,33 @@ public Client(Configuration configuration)
...
+                    };
 
+                    HttpResponseMessage httpResponseMessage = await client.SendAsync(httpRequestMessage);
+                    response = await GetResponse<Token>(httpResponseMessage);
+                }
+                if (response.StatusCode < 500)
+                {
+                    return response;
+                }
+                else if (this.Configuration.InternalServerErrorRetryInterval > 0)
```
Comment (hajoeichler) [inline commercetools.NET/Common/Client.cs]: This check should be done only once during construction of the Configuration object.

**Comment by purquhart** [comment]
> Those number should not be hardcoded, but configurable.
- Please note the SDK does not use configuration files so default values are necessary. These numbers are configurable when you utilize the optional parameters of the Configuration class constructor, and when you don't, it utilizes the default values. I have set the default of those numbers to be 1 retry with a 100ms delay interval; would you rather that the default be 0 and 0?

> Could you not abstract the actual call so you don't have to repeat the implementation 5 times.
- Have one call method that can be configured to be POST, DELETE etc and have am optional wrapper for repetition.
We can abstract the call if you wish, however the reason why these calls were separated in the original code was to reduce complexity towards any refactoring that was specific to the calls since they utilize different parameters beyond the httpmethod - would you please confirm you would still like this to be abstracted in the manner you are suggesting?

> There should be a max retry < 10 (better 3 or 5). This constraint should be checked in the code.
- What would you like the max retry to be, and what is the desired outcome - if they set a value that is higher than the limit should we automatically set it to the limit, or should we deliver an exception? 

> This check should be done only once during construction of the Configuration object.
- I'm unsure what is actually being asked for here as the only time it is checked is when an applicable error has occured, and could mean a variety of things since the Configuration object is passed into the constructor of the Client object - could you clarify further? Do you perhaps mean that the max retry that is being requested above should only be checked once?

File: commercetools.NET/Common/Client.cs [hunk 2]
```diff
@@ -82,22 +83,32 @@ public Client(Configuration configuration)
...
+                    HttpResponseMessage httpResponseMessage = await client.SendAsync(httpRequestMessage);
+                    response = await GetResponse<T>(httpResponseMessage);
+                }
+                if (response.StatusCode < 500)
+                {
+                    return response;
+                }
+                else if (this.Configuration.InternalServerErrorRetryInterval > 0)
+                {
+                    Thread.Sleep(this.Configuration.InternalServerErrorRetryInterval);
```
Comment (jenschude) [inline commercetools.NET/Common/Client.cs]: Unsure if the Thread.Sleep is correct at this point, as Thread.Sleep stops the whole thread execution. An ```await Task.Delay(this.Configuration.InternalServerErrorRetryInterval);``` would just stop the execution of the task for the specified amount of time and is more inline with the async/await modelling.

File: commercetools.NET/Common/Client.cs (hunk 1 above)
Comment (jenschude) [reply commercetools.NET/Common/Client.cs]: No. As the application may run longer than the token life time the refreshToken task may be called multiple times.

**Comment by purquhart** [comment]
Thank you @hajoeichler and @jenschude for your comments - I've applied the Task.Delay refactoring to the Client class as suggested and committed it, excellent idea.

File: commercetools.NET/Common/Client.cs [hunk 3]
```diff
@@ -10,6 +10,7 @@
 
 using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
+using System.Threading;
```
Comment (jenschude) [inline commercetools.NET/Common/Client.cs]: Is this import still needed after the Thread.sleep change? Else everything seems to be okay.

Comment (purquhart) [reply commercetools.NET/Common/Client.cs]: Thanks @jenschude, removed the reference and committed

**Pull request merged**